text/2442-simple-postfix-macros.md

```diff
@@ -0,0 +1,394 @@
+
+The approach proposed in this RFC does not preclude specifying a richer system
+in the future; such a future system could use a new designator other than
+`self`, or could easily extend this syntax to add further qualifiers on `self`
+(for instance, `$self:self:another_designator` or `$self:self(argument)`).
+
+Rather than writing `expr.macroname!()`, we could write `expr.!macroname()` or
+similar, placing the `!` closer to the `.` of the method call. This would place
+greater attention on the invocation, but would break the similarity with
+existing macro naming that people have grown accustomed to spotting when
+reading code. This also seems more likely to get confused with the prefix unary
+`!` operator.
+
+In the syntax to define a postfix macro, we could use just `$self` rather than
+`$self:self`. `$self` is not currently valid syntax, so we could use it for
+this purpose without affecting any existing valid macro. This would make such
+declarations look closer to a method declaration, which uses `self` without a
+type. However, macros do currently allow `self` as the name of a macro argument
+when used with a designator, such as `$self:expr`; this could lead to potential
+confusion, and would preclude some approaches for future extension.
+
+In the syntax to define a postfix macro, rather than using
+```
+($self:self, $arg:expr) => (...)
+```
+we could use
+```
+$self:self.($arg:expr) => (...)
+```
+. This would have the advantage of looking more similar to the invocation, but
+the disadvantage of looking much different than method definition syntax (in
+which `self` appears within the function arguments).
+
+In the syntax to define a postfix macro, rather than using
+```
+($self:self, $arg:expr) => (...)
+```
+we could use
+```
+($self:self. $arg:expr) => (...)
+```
+or
+```
+($self:self $arg:expr) => (...)
+```
+. Using a `.` might be evocative of method syntax, but would be unusual and
+harder to remember. Using no delimiter at all would reflect that the `,` does
+not actually match a literal `,` in syntax, and might be clearer when using
+macros whose arguments use unusual syntax substantially different than method
+arguments (e.g. `expr.mac!(:thwack => boing | :poit <= narf)`) but would be
+confusing and error-prone for macros intended to resemble method calls (e.g.
+`expr.mac!(arg1, arg2, arg3)`). (Making the `,` optional seems even more
+confusing and prone to ambiguity.)
+
+We could omit the `k#autoref` mechanism and only support `self`. However, this
+would make `some_struct.field.postfix!()` move out of `field`, which would make
+it much less usable.
+
+We could omit the `k#autoref` mechanism in favor of requiring the macro to
+specify whether it accepts `self`, `&self`, or `&mut self`. However, this would
+prevent writing macros that can accept either a reference or a value, violating
+user expectations compared to method calls (which can accept `&self` but still
+get called with a non-reference receiver).
```

Comment:
Why can't we just use `&$self` as an expression and have that evaluate to the tokens that were passed as the `self`? Just like with other arguments, if you don't want it evaluated twice, first evaluate it into a let binding, then use that twice.